src/openzeppelin/utils/structs/enumerableset.cairo

```diff
@@ -0,0 +1,141 @@
+            syscall_ptr : felt*, 
+            pedersen_ptr : HashBuiltin*, 
+            range_check_ptr
+        }(set_id : felt, value : felt) -> (contains : felt):
+        let (index) = EnumerableSet_indexes.read(set_id, value)
+        if index == 0:
+            return (FALSE)
+        end
+        return (TRUE)
+    end
+
+    func add{
+            syscall_ptr : felt*, 
+            pedersen_ptr : HashBuiltin*, 
+            range_check_ptr
+        }(set_id : felt, value : felt) -> (success : felt):
+        alloc_locals
+        let (contains) = EnumerableSet.contains(set_id, value)
+
+        if contains == FALSE:
+            let (size) = EnumerableSet_sizes.read(set_id)
+            local new_size = size + 1
+            EnumerableSet_values.write(set_id, new_size, value)
+            EnumerableSet_indexes.write(set_id, value, new_size)
+            EnumerableSet_sizes.write(set_id, new_size)
+            return (TRUE)
+        end
+
+        return (FALSE)
+    end
+
+    func remove{
+            syscall_ptr : felt*, 
+            pedersen_ptr : HashBuiltin*, 
+            range_check_ptr
+        }(set_id : felt, value : felt) -> (success : felt):
+        alloc_locals
+        let (value_index) = EnumerableSet_indexes.read(set_id, value)
+
+        if value_index == 0:
+            return (FALSE)
+        end
+
+        let (last_index) = EnumerableSet_sizes.read(set_id)
+        local res = last_index - value_index
+        local new_last_index = last_index - 1
+
+        let (last_value) = EnumerableSet_values.read(set_id, last_index)
+
+        EnumerableSet_values.write(set_id, last_index, 0)
+        EnumerableSet_indexes.write(set_id, value, 0)
+        EnumerableSet_sizes.write(set_id, new_last_index)
+
+        # if removed value index is not last
+        if res != 0:
+            EnumerableSet_values.write(set_id, value_index, last_value)
+            EnumerableSet_indexes.write(set_id, last_value, value_index)
+        end
+
+        return (TRUE)
+    end
+
+    func length{
+            syscall_ptr : felt*, 
+            pedersen_ptr : HashBuiltin*, 
+            range_check_ptr
+        }(set_id : felt) -> (length : felt):
+        let (size) = EnumerableSet_sizes.read(set_id)
+        return (size)
+    end
+
+    func at{
+            syscall_ptr : felt*, 
+            pedersen_ptr : HashBuiltin*, 
+            range_check_ptr
+        }(set_id : felt, index : felt) -> (value : felt):
+        let (value) = EnumerableSet_values.read(set_id, index)
+        return (value)
+    end
+
+    func values{
+            syscall_ptr : felt*, 
+            pedersen_ptr : HashBuiltin*, 
+            range_check_ptr
+        }(set_id : felt) -> (res : felt*):
+        let index = 1
+        let values : felt* = alloc()
+        let (size) = EnumerableSet_sizes.read(set_id)
+        let size_oob = size + 1
+        let (vals) = _values_helper(
+            set_id, index, size_oob, values)
+        return (res=vals)
```

Comment:
yea but that would mess up the index mapping since we test for the index being 0 as it not being part of the set (same behavior thats in the solidity version of the code)

Reply:
right, I missed line 132, that's the one that breaks what I just said. We could then `+1` there, wouldn't that remove every other offset?

Reply:
possibly but i dont know if its that big of a gain to switch around